### DISCUSSION TITLE
Feature Request: Option to bypass audio processing and maintain original sample rate/bit depth

### DISCUSSION THREAD

**Discussion by epochaudio** [music-assistant/support#3327] [Feature requests and ideas]
Title: Feature Request: Option to bypass audio processing and maintain original sample rate/bit depth

## Description
I would like to request a feature to bypass all audio processing in Music Assistant and output the audio stream with its original sample rate and bit depth. This would allow for bit-perfect playback without any resampling or audio manipulation.

## Use Case
- For users with high-end audio equipment who want to preserve the original audio quality
- For streaming high-resolution audio files without any downsampling
- For maintaining bit-perfect playback chain from source to output device

## Proposed Solution
Add a new option in the stream controller settings:
- [x] Add a "Bypass Processing" toggle in core settings
- [ ] When enabled:
  - Disable resampling
  - Disable volume normalization
  - Disable any audio effects/processing
  - Pass through the original sample rate
  - Maintain original bit depth
  - Keep original audio format (e.g., FLAC, DSD)

## Technical Details
The implementation would require:
1. Identifying where the audio processing pipeline is in the codebase
2. Adding a bypass mode that skips the processing chain
3. Ensuring the original audio parameters are preserved throughout the streaming process
4. Adding appropriate UI controls in the settings

## Benefits
- Better audio quality for audiophile setups
- Lower CPU usage (no processing overhead)
- True bit-perfect playback capability
- Support for high-resolution audio formats at native quality

## Additional Context
This feature would be particularly useful for users who:
- Have DACs that handle their own processing
- Want to ensure no quality loss in their audio chain
- Need bit-perfect playback for specific use cases

Would this be something that could be considered for implementation?

**Comment by marcelveldt** [comment]
This is not the intended audience/usecase for Music Assistant.

Strictly taken we do close to no processing as soon as you disable volume normalization. The only thing we do is decode into PCM internally and that has to do with the fact we need to be compatible with a whole bunch of audio sources and target players - there is no loss of audio data at all. 

Bits are bits. Audio quality is lost when you encode into a lossy format (or output to some shitty speakers), not in other ways.
Audiophiles that believe in these kind of myths can resort to software like Roon. We are more targeted to a broader audience.
Bit perfect is just marketing bullshit. Sorry to say it that way. I'm a believer of common sense and thats why/how I've built this all (and got me very far).

**Reply by yvmexcl** [reply]
It would save a lot of CPU cycles, I have problems with qbuz to play from 96khz 24bit and I think the limitation is the Rpi4, either in the same HA rpi4 or a separate rpi4, with other software that allows bitperfect I don't have it. Beyond marketing, it is being practical in the use of the limited CPU in the rpi

**Reply by marcelveldt** [reply]
That is actually another issue which has been fixed in the beta. We will provide a new stable patch soon

**Reply by z0mbieprocess** [reply]
While I agree with you @marcelveldt it's still a bummer my DAC LEDs do not turn cyan when playing DSD files.

**Comment by YannChris** [comment]
Hi, 
In the settings of the player, I unchecked all options related to sound and selected all bitrates supported by the target speaker.
When playing files coming from a ripped CD, I notice:
- Input Flac 44,1 kHz / 16 bits
- Output Flac 96 kHz / 24 bits
I doubt that this upsampling will improve the sound, but it stresses the processor (in my case, MA and HA are run in a virtual machine installed in a NAS).
That's why I agree with the request and would appreciate to have a "bypass audio processing" option.
Regards,
Y.

**Reply by Lipown** [reply]
Yes please. I just noticed the same...

**Reply by OzGav** [reply]
We have closed this because the people posting misunderstand how this works. Did you read Marcel's reply?

**Reply by Lipown** [reply]
@OzGav yes I read that but still as @yvmexcl mentioned selecting all supported bitrates of my receiver making Music Assistant converting all streams to the highest bitrate setting. This making the process quite resource hungry. It would save a lot if I can do passthrough only...

**Reply by OzGav** [reply]
I would say if a RPi 4 can keep up with the stream, notwithstanding any spike in resource usage, then the project is meeting its intended objectives. If you are using a lesser processor than a RPi4 then you are probably out of scope. If you are using a RPi4 or equivalent and it is not able to keep up with the stream demands the you can open an issue about that.

**Comment by TarcisioMenezes** [comment]
Hi guys, 
My question is , if the Output codec for my player is FLAC (and audio processing is OFF, i.e normalization, limiter, DSP and etc), and my source file is also FLAC (with sample rate and bit depth compatible with the player), why to decode and encode it again? 

On Marcel's post he said "The only thing we do is decode into PCM internally and that has to do with the fact we need to be compatible with a whole bunch of audio sources and target players." 

in other words , why to not send the file the way it is if the player is already compatible with the source file and there is no audio processing set for that player? is it really necessary to decode and encode the audio if there is no change to be made?

It's not about being able to play bit-perfect (that's not a dealbreaker for me honestly), it's about doing unnecessary steps and simplify the pipeline when possible.

Just to be clear, I'm not here to say what should or not should be. I've been using Music Assistant for 3 weeks now, and I love it! 
I'm just wondering if that's really necessary, and if it is, why? 

Thank you guys!

E.g
<img width="366" height="298" alt="image" src="https://github.com/user-attachments/assets/10933141-ddf2-4de2-b9bd-326f6557b0fa" />

**Reply by OzGav** [reply]
Because MA supports so many different players with many different capabilities, and due to the available functions (which you have listed that you have disabled) this all means that processing is likely required in most cases and in order to ensure no information is lost the files are converted to 32bit PCM.

**Reply by TarcisioMenezes** [reply]
Hey OzGav, thanks for the reply. 

I understand that decoding and encoding is necessary when there is some kind of manipulation/audio processing or when the provider's file uses a different codec then the 'output codec' for the player but my question is:

if the music provider tracks properties (codec, bit depth, sample rate) matches with the player  AND the audio processing functions/features (limiter, normalizer, eq and etc...) are turned off. What's the point of decoding and encode the file?

Couldn't be something like: " Ok, audio processing is off and track matches with the player capabilities. Let's send the track the way it is cause it will be compatible. " 

I hope my question makes sense, if not, can you please explain me how does it work?
I appreciate it !

Thank you

**Reply by OzGav** [reply]
It does but its a lot of work for not a lot of gain (indeed no gain that I can see) so the core team will not be progressing this. Sorry.

**Reply by unaszplodr** [reply]
> It does but its a lot of work for not a lot of gain (indeed no gain that I can see) so the core team will not be progressing this. Sorry.

The gain can be measured in free cpu cycles and reduced power consumption. For example, my Celeron N2830 sits at around 0.80 load average (1 min) for bit-perfect playback. It is not fair to state there would be no benefit to a passthrough mode. That is simply, patently not true. How many users would benefit from the additional headroom reclaimed when unnecessary processing is bypassed is less obvious — I don't have access to usage metrics, ofc.

**Comment by maisun** [comment]
For the sake of understanding - I play Flacs in various sampling rate, I notice output has the same sampling rate but always upscaled to 24bits. Anybody know how to disable that? 
BTW: I also agree if the player is compatible with the format it should not decode and encode again.
It potentially gives issue to MQA stream from Tidal (not so big deal nowadays).
